### ISSUE TITLE
shadowJar exclude nebula provided dependencies

### ISSUE THREAD

**Issue by kzarzycki** [johnrengelman/shadow#166]
Hi there, 
I would like to exclude all dependencies that I marked as provided, i.e. added to "provided" configuration, that comes with nebula plugin: (https://github.com/nebula-plugins/gradle-extra-configurations-plugin)

I've tried for long time but failed... How can I exclude not a single dependency but the whole configuration? 
I've tried sth like this: 

```
shadowJar {
    dependencies {
        // approach 1: FAIL
        exclude project.configurations.provided
        // approach 2: FAIL
        project.configurations.provided.dependencies.each {
            def dep = "${it.group}:${it.name}:${it.version}"
            exclude(dependency(dep))
        }
   }
}
```

Nebula plugins are pretty standard for gradle projects, it would be nice if they could work together...

What would is the other way, with shadowJar, to exclude dependencies, but not by repeating all excluded ones in the shadowJar task? 

Thanks
Krzysztof

**Comment by johnrengelman** [comment]
I actually see this a problem with Nebula's provided plugin. The issue here in my mind is that they are declaring the `provided` configuration as a root of `compile` (https://github.com/nebula-plugins/gradle-extra-configurations-plugin/blob/master/src/main/groovy/nebula/plugin/extraconfigurations/ProvidedBasePlugin.groovy#L60). IMO that's not appropriate, but to each their own.
Anyway, this causes the contents of `provided` to be inherited into `runtime` which is then merged into the shadow Jar. I'm surprised your "approach 2" doesn't work. 
In theory, that could even be simplified to:
```
project.configurations.provided.dependencies.each {
  exclude(it)
}
```
I'll have to add a test for this, but also perhaps a helper to exclude an entire configuration.

**Comment by ncwoehler** [comment]
We faced the same problem as we also use the Nebula provided plugin. 
I think the problem with approach two is that you do not exclude transitive dependencies of the provided configuration. Our solution:

```
// ensure provided dependencies are not compiled into shadowJar
def firstLevelProvided = project.configurations.provided.getResolvedConfiguration().getFirstLevelModuleDependencies()
def artifactsToExclude = getResolvedArtifacts(firstLevelProvided)

artifactsToExclude.each { artifact ->
    project.logger.info "Excluding ${artifact} from being bundled into the shadow jar."
    shadowJar {
        dependencies {
            exclude(dependency(artifact))
        }
    }
}

def getResolvedArtifacts(Set<ResolvedArtifact> artifacts) {
    Set<String> resolvedArtifacts = [] as Set
    artifacts.each {
        // add current artifact
        resolvedArtifacts << "${it.moduleGroup}:${it.moduleName}:${it.moduleVersion}"

        // recursion to add children
        resolvedArtifacts += getResolvedArtifacts(it.children)
    }
    return resolvedArtifacts
}
```

**Comment by johnrengelman** [comment]
Yeah, actually everything proposed can have issues in some scenarios. Say for example that you have a transitive dependency that gets resolved in both the `provided` and `compile` configurations (assuming that `compile` is isolated from `provided`), does it get included or excluded then from being merged into shadow? In the solution above, it would be excluded, but I don't believe that's the correct answer. 

You could maybe solve this by creating another configuration that duplicates the entries in `compile` and `runtime` without inheriting from them. Then that configuration would be isolated from the `provided` configuration and you could modify Shadow to use it instead:
```
shadowJar {
  configurations = [ project.configurations.isolated ]
}
```

Again, IMO this is an issue with Nebula's implementation of `provided`. While they are declaring that `provided` is a root of `compile`, there's nothing we can do to solve this 100%.
